client/dashboard/leaderboards/index.js

```diff
@@ -0,0 +1,184 @@
+/** @format */
+/**
+ * External dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { Component, Fragment } from '@wordpress/element';
+import { compose } from '@wordpress/compose';
+import { isEqual } from 'lodash';
+import PropTypes from 'prop-types';
+import { SelectControl, ToggleControl } from '@wordpress/components';
+import { withDispatch } from '@wordpress/data';
+
+/**
+ * WooCommerce dependencies
+ */
+import { EllipsisMenu, MenuItem, MenuTitle, SectionHeader } from '@woocommerce/components';
+
+/**
+ * Internal dependencies
+ */
+import withSelect from 'wc-api/with-select';
+import TopSellingProducts from './top-selling-products';
+import './style.scss';
+
+class Leaderboards extends Component {
+	constructor( props ) {
+		super( ...arguments );
+		this.state = {
+			hiddenLeaderboardKeys: props.userPrefLeaderboards || [],
+			rowsPerTable: props.userPrefLeaderboardRows || 5,
+		};
+
+		this.toggle = this.toggle.bind( this );
+	}
+
+	componentDidUpdate( {
+		userPrefLeaderboardRows: prevUserPrefLeaderboardRows,
+		userPrefLeaderboards: prevUserPrefLeaderboards,
+	} ) {
+		const { userPrefLeaderboardRows, userPrefLeaderboards } = this.props;
+		if ( userPrefLeaderboards && ! isEqual( userPrefLeaderboards, prevUserPrefLeaderboards ) ) {
+			/* eslint-disable react/no-did-update-set-state */
+			this.setState( {
+				hiddenLeaderboardKeys: userPrefLeaderboards,
+			} );
+			/* eslint-enable react/no-did-update-set-state */
+		}
+		if (
+			userPrefLeaderboardRows &&
+			parseInt( userPrefLeaderboardRows ) !== parseInt( prevUserPrefLeaderboardRows )
+		) {
+			/* eslint-disable react/no-did-update-set-state */
+			this.setState( {
+				rowsPerTable: parseInt( userPrefLeaderboardRows ),
+			} );
+			/* eslint-enable react/no-did-update-set-state */
+		}
+	}
```

Comment:
This seems really complex to effectively copy props into state (and the `eslint-disable`s smell)..

Is there a different way to manage the components state and still perform the side effects of updating user preferences without mapping to props? Perhaps trigger a `getCurrentUserData()` inside `componentDidMount()` and write to state there?

Reply:
While I agree that it seems overly complex, I think this is the [accepted method of handling this](https://reactjs.org/docs/react-component.html#componentdidupdate) and how we do so in other components. Unfortunately, we can’t set this in `componentDidMount()` since user data is fetched asynchronously and would not get set in time. 

If you have other suggestions, I’m open to trying them.

Reply:
Seems like this will be a non-issue when the initial state fetch is refactored.